midterm/ProductDetails.tsx

```diff
@@ -0,0 +1,218 @@
+import { useNavigation, useRoute } from "@react-navigation/native";
+import { useEffect, useRef, useState } from "react";
+import { Text, View } from "react-native";
+import { TouchableOpacity, StyleSheet, Image } from "react-native";
+import Ionicons from "@expo/vector-icons/Ionicons";
+import { useAsyncStorage } from "@react-native-async-storage/async-storage";
+import { useFetch } from "./hooks/useFetch";
+import { SafeAreaView } from "react-native-safe-area-context";
+import { favoriteItem } from "./types/favoriteItem";
+
+type RouteParams = {
+	url: string;
+};
+
+export const ProductDetails = () => {
+	const buttonPressed = useRef(false);
+	const { params }: any = useRoute();
+	const typedParams: RouteParams = params;
+
+	const [favoritesList, setFavoritesList] = useState<favoriteItem[]>([]);
+
+	const [inFavorites, setInFavorites] = useState(false);
+
+	const { getItem, setItem } = useAsyncStorage("@storage_key");
+
+	const { data, loading, error } = useFetch(typedParams.url);
+
+	const heartButton = () => {
+		return (
+			<View style={styles.heartButtonContainer}>
+				<TouchableOpacity
+					onPress={() => {
+						handleHeartPressed();
+					}}
+				>
+					<Ionicons
+						name={inFavorites ? "heart" : "heart-outline"}
+						size={30}
+						color={"indianred"}
+					/>
+				</TouchableOpacity>
+			</View>
+		);
+	};
+
+	const navigation = useNavigation();
+	// navigation.setOptions({
+	//     headerRight: heartButton,
+	//     headerTitle: "Product Details"
+	// })
+
+	useEffect(() => {
+		navigation.setOptions({
+			headerRight: heartButton,
+			headerTitle: "Product Details",
+		});
+	}, [inFavorites, navigation]);
+
+	let zoomedInData: any;
+
+	if (data !== null) {
+		zoomedInData = data.data;
+	}
+
+	useEffect(() => {
+		retrieveFavorites();
+	}, []);
+
+	useEffect(() => {
+		const searchResult = favoritesList.findIndex((favorite) => {
+			return favorite.url === typedParams.url;
+		});
+
+		if (searchResult !== -1) {
+			setInFavorites(true);
+		}
+	}, [favoritesList, typedParams.url]);
+
+	useEffect(() => {
+		if (buttonPressed.current) {
```

Comment:
I'm not sure why you would need `useRef` here. We know the button is pressed once `handleHeartPressed ` is invoked. All of this logic should be part of `handleHeartPressed` and not in `useEffect`